File: mass_scheduler.py
#!/usr/bin/env python3

import re
import sys
import time

from argparse import ArgumentParser
from collections import ChainMap
from termcolor import colored

# List of attributes to parse => push onto FIFO queue
attributes = [
    'current_state',
    'host_name',
    'plugin_output',
    'problem_has_been_acknowledged',
    'scheduled_downtime_depth',
    'service_description',
]

# Precompile Regular expression to match lines
attributes_regex = re.compile("|".join(attributes))


def parse_args(args):
    parser = ArgumentParser()
    parser.add_argument('-u', '--username')
    parser.add_argument('-m', '--msg')
    parser.add_argument('-r', '--regex')
    return parser.parse_args(args)


def parse_missing_args(args):
    """Attempts to fetch missing arguments by prompting user"""

    while not args.username:
        args.username = get_input('Enter your Username: ')
    while not args.msg:
        args.msg = get_input('Enter message used for acknowledgement:')
    while not args.regex:
        args.regex = get_input('Enter desired search regex:')

    # Pre-Compile the regular expression
    args.regex = re.compile(r'{}'.format(args.regex))
    return args


def get_file(fname, flags):
    """Attempts to open a file on the system"""
    try:
        f = open(fname, flags)
    except (OSError, IOError) as error:
        raise SystemExit(error)
    else:
        return f


def get_input(prompt, choices={}):
    """Provides user with input prompt, quits on escape key-code"""
    try:
        input_string = input(prompt + '\n-> ')
    except (EOFError, KeyboardInterrupt):
        raise SystemExit(colored("\nQuitting!", 'red'))
    else:
        # Retry if input_string is empty
        return input_string or get_input(colored('Retry', 'red'), choices)


def sanitize(text):
    """Removes whitespace and lowercases text"""
    return text.strip().lower()


def is_interesting(service):
    """Should the service be available for acknowledging?"""

    # If the service state is OK, we don't care
    if int(service['current_state']) == 0:
        return False

    # Skip if already acknowledged
    if int(service['problem_has_been_acknowledged']) != 0:
        return False

    # Skip if currently in scheduled downtime
    if int(service['scheduled_downtime_depth']) != 0:
        return False

    return True

def prompt_action(service):
    """Prompts the user for input for each pending service"""

    # Display service description and hostname
    service_info = "{service_description} @ {host_name}".format(**service)
    print(colored('Service:', 'yellow'), service_info)

    # Display plugin output
    output_info = "Output: {plugin_output}".format(**service)
    print(colored('Output:', 'cyan'), output_info)

    # Prompt user for [y/n/s]
    choices, colours = ('y', 'n', 's'), ('green', 'red', 'blue')
    text = (colored(t, c) for (t, c) in zip(choices, colours))
    choice = sanitize(get_input('[{}/{}/{}]'.format(*text)))

    # Retry if the choice is invalid
    return choice if choice in choices else prompt_action(service)


def get_date(prompt):
    while True:
        try:
            return int(get_input(prompt))
        except ValueError:
            print(colored('Invalid date', 'red'))


def handle_choice(choice, fifo_queue, service, args):
    now = int(time.time())

    # If the user wishes to Schedule Downtime
    if choice.startswith('s'):
        duration = get_date('Duration [mins from now]')

        dates = {
            'start_time': now,
            'duration': duration * 60,
            'end_time': now + duration * 60,
        }
        data = ChainMap(service, vars(args), dates)

        write_data = "[{start_time}] SCHEDULE_SVC_DOWNTIME;{host_name};" \
            "{service_description};{start_time};{end_time};1;0;{duration};" \
            "{username};{msg}\n"

        fifo_queue.write(write_data.format(**data))

    # If the user wishes to Acknowledge problem
    elif choice.startswith('y'):

        write_data = "[{time}] ACKNOWLEDGE_SVC_PROBLEM;{host_name};" \
            "{service_description};1;0;0;{username};{msg}\n"

        data = ChainMap(service, vars(args), {'time': now})
        fifo_queue.write(write_data.format(**data))


def main():
    status_file = get_file('status.dat', 'r')
    fifo_queue = get_file('outfile.txt', 'w')

    args = parse_missing_args(parse_args(sys.argv[1:]))

    # Store attributes in a dictionary
    service = {}

    # Read in lines into a generator object
    lines = (line.strip() for line in status_file)

    is_service = False

    for line in lines:

        # Skip comments and blank lines
        if line.startswith('#') or line == '':
            continue

        # Reached the start of a 'servicestatus' block
        elif line.startswith('servicestatus') and line.endswith('{'):
            is_service = True

        # Reached the end of a 'servicestatus' block
        elif line.startswith('}'):

            # Acknowledge previously parsed service
            if is_service:

                # Only proceed if current service matches search term
                desc = service.get('service_description', False)

                if is_interesting(service) and args.regex.match(desc):

                    choice = prompt_action(service)
                    handle_choice(choice, fifo_queue, service, args)

            # Delete attributes set from previous service
            service.clear()
            is_service = False

        if is_service:
            match = re.match(attributes_regex, line)
            if match:
                key = match.group()
                service[key] = line.split('=', maxsplit=1)[-1]

    print(colored('Finished!', 'green'))


if __name__ == '__main__':
    main()

File: mass_scheduler.pl
#!/usr/bin/perl

###########################################################################################
#   This script provides help to acknowledge multiple services during a large maintenance
#   Sometimes host groups and service groups do not suffice
#      Script reads nrpe.cfg for the location of the status.dat and the FIFO file
#      Script should be able to write to the FIFO file
#      Command is run interactively
#      Santiago Velasco - sanxiago.com
#########################################################################################

$nagios_config = '/etc/nagios3/nagios.cfg';

eval {
    local $SIG{ALRM} = sub { die "timeout" };
    alarm 600;
    $input = main();
    alarm 0;
    exit 0;
};
if ($@ and $@ =~ /timeout/) {
    print "Run time out\n";
    close(STATUS);
    die ;
} elsif ($@) {
    close (STATUS);
    die "Other error: $@\n";
}

sub main {
    my $command_file;
    my $status_file;
    if (-r $nagios_config){
        open (CONFIG, $nagios_config);
        while (<CONFIG>){
            $status_file = ($_=~/status_file=(.*)/)?$1:$status_file;
            $command_file = ($_=~/command_file=(.*)/)?$1:$command_file;
        }
        close(CONFIG);
    }
    else {
        print STDERR "FAILED TO READ NAGIOS CONFIG FILE\n";
        exit 1;
    }
    my $time = time();
    my %state = (1 ,'WARNING', 2,'CRITICAL', 3,'UNKNOWN');
    my $user = $ARGV[0];
    my $msg = $ARGV[1];
    my $search_string = $ARGV[2];



    print STDERR "\n\nACKNOWLEDGE AND SCHEDULE DOWNTIME FOR MULTIPLE SERVICES\n\n";

    while(!defined($user) or $user =~ /\;|\[|\]/  or length($user)<=1){
        print STDERR "Type in yout USER that acknowledges:\n";
        $user = <>;
        $user =~ s/\n//;
    }
    while (!defined($msg) or $msg =~ /\;|\[|\]/ or length($msg)<=1 ){
        print STDERR "Type in the MESSAGE that will be used for all acknowledges:\n";
        $msg = <>;
        $msg =~ s/\n//;
    }
    print STDERR "Type in a string that matches the service_description of the services you want to ack.\n Leave it blank to list all alerts):\n";
    $search_string = <>;
    $search_string =~ s/\n//;
    if(length($search_string)<=1){
        $search_string='.*';
    }

    if (-r $status_file){
        open (STATUS, $status_file);
    }
    else {
        print STDERR "FAILED TO READ NAGIOS STATUS FILE\n";
        exit 1;
    }
    while(<STATUS>){
        if($_ =~ /(service|servicestatus) \{/){
            $is_service = 1;
        }
        if($_ =~ /\}/ and $service_description=~/$search_string/){
            $is_service =0;
            if(defined($current_state) and $current_state and $acknowledged==0 and $scheduled_downtime==0 ){
                # Command Format:
                # [time] ACKNOWLEDGE_SVC_PROBLEM;<host_name>;<service_description>;<sticky>;<notify>;<persistent>;<author>;<comment>
                # [time] SCHEDULE_SVC_DOWNTIME;<host_name>;<service_desription><start_time>;<end_time>;<fixed>;<trigger_id>;<duration>;<author>;<comment>
                undef($ack_true);
                print STDERR "\n---------------------------------------------------------\n";
                print STDERR "Acknowledge $service_description @ $host_name $current_state".$state{$current_state}."?\n$plugin_output\n[y/n/s] (s followed by the number of minutes of scheduled downtime) (Enter to skip)\n";
                $ack_true=<>;
                # if acknowledge yes
                if($ack_true=~/^y/){
                    if (!(-w $command_file)){ print STDERR "FAILED TO OPEN FIFO FILE"; exit 1; }
                    open (CMD, '>>'.$command_file);
                    print CMD "[$time] ACKNOWLEDGE_SVC_PROBLEM;$host_name;$service_description;1;0;0;$user;$msg\n";
                    close (CMD);
                    # if schedule downtime
                }elsif($ack_true=~/^s(.*)/){
                    my $duration = $1;
                    if($duration=~/[^\d]*([0-9]+).*/){
                        #expect duration in minutes convert to seconds
                        $duration=int($1)*60;
                    }else{
                        $duration=3600;
                    }
                    my $end_time = $time + $duration;

                    if (!(-w $command_file)){ print STDERR "FAILED TO OPEN FIFO FILE"; exit 1; }
                    open (CMD, '>>'.$command_file);
                    print CMD "[$time] SCHEDULE_SVC_DOWNTIME;$host_name;$service_description;$time;$end_time;1;0;$duration;$user;$msg\n";
                    close (CMD);
                }
            }
            undef($current_state);
            undef($host_name);
        }
        if($is_service){
            if($_=~/host_name\=(.*)/){
                $host_name=$1;
            }
            if($_=~/service_description\=(.*)/){
                $service_description=$1;
            }
            if($_=~/current_state\=([0-9]*)/){
                $current_state=$1;
            }
            if($_=~/problem_has_been_acknowledged\=([0-9]*)/){
                $acknowledged=$1;
            }
            if($_=~/plugin_output\=(.*)/){
                $plugin_output=$1;
            }
            if($_=~/scheduled_downtime_depth\=([0-9]*)/){
                $scheduled_downtime=$1;
            }
        }
    }
    close(STATUS);
}
